apiserver/internal/handlers/resource/resource.go

```diff
@@ -0,0 +1,417 @@
+		}
+	case "PUT":
+		if err := h.changeAllowedFunc(req.Context()); err != nil {
+			if err := sendError(resp, err, h.logger); err != nil {
+				h.logger.Errorf("returning error to user: %v", err)
+			}
+			return
+		}
+		response, err := h.upload(resourceService, req, tagToUsername(tag))
+		if err != nil {
+			if err := sendError(resp, err, h.logger); err != nil {
+				h.logger.Errorf("returning error to user: %v", err)
+			}
+			return
+		}
+		if err := sendStatusAndJSON(resp, http.StatusOK, &response); err != nil {
+			h.logger.Errorf("sending response: %v", err)
+		}
+	default:
+		if err := sendError(resp, jujuerrors.MethodNotAllowedf("unsupported method: %q", req.Method), h.logger); err != nil {
+			h.logger.Errorf("returning error to user: %v", err)
+		}
+	}
+}
+
+func (h *ResourceHandler) download(service ResourceService, req *http.Request) (io.ReadCloser, int64, error) {
+	defer func() {
+		if req.Body != nil {
+			req.Body.Close()
+		}
+	}()
+
+	query := req.URL.Query()
+	application := query.Get(":application")
+	name := query.Get(":resource")
+
+	uuid, err := service.GetResourceUUIDByApplicationAndResourceName(req.Context(), application, name)
+	if errors.Is(err, resourceerrors.ResourceNotFound) {
+		return nil, 0, jujuerrors.NotFoundf("resource %s of application %s", name, application)
+	} else if err != nil {
+		return nil, 0, fmt.Errorf("getting resource uuid: %w", err)
+	}
+
+	res, reader, err := service.OpenResource(req.Context(), uuid)
+	if errors.Is(err, resourceerrors.StoredResourceNotFound) {
+		return nil, 0, jujuerrors.NotFoundf("resource %s of application %s has no blob downloaded on controller", name, application)
+	} else if err != nil {
+		return nil, 0, errors.Errorf("openeing resource %s for application %s: %w", name, application, err)
+	}
+	return reader, res.Size, err
+}
+
+func (h *ResourceHandler) upload(resourceService ResourceService, req *http.Request, username string) (_ *params.UploadResult, err error) {
+
+	reader, uploaded, err := h.readResource(resourceService, req)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	defer func() {
+		if err != nil {
+			reader.Close()
+		}
+	}()
+
+	err = resourceService.StoreResourceAndIncrementCharmModifiedVersion(
+		req.Context(),
+		resource.StoreResourceArgs{
+			ResourceUUID:    uploaded.UUID,
+			Reader:          reader,
+			RetrievedBy:     username,
+			RetrievedByType: resource.User,
+			Size:            uploaded.Size,
+			Fingerprint:     uploaded.Fingerprint,
+			Origin:          charmresource.OriginUpload,
+			Revision:        -1,
+		},
+	)
+	if err != nil {
+		return nil, errors.Errorf("storing resource %s of application %s: %w", uploaded.Resource.Name, uploaded.Application, err)
+	}
+
+	res, err := resourceService.GetResource(req.Context(), uploaded.UUID)
+	if err != nil {
+		h.logger.Errorf("getting uploaded resource details: %w", err)
+		return &params.UploadResult{}, nil
+	}
```

Comment:
This doesn't look right at all. If you can't get the resource, I'd want to return an error. If you can't find the resource and you did an upload, you get a 200?

Reply:
Good point, this should be an error, the resource should be in the DB here so if we get an error something is wrong. The reasoning here was that the resource has successfully been uploaded in the store resource request and this subsequent check is just to get the data to return about it, so even if it fails, the resource has still be uploaded.